File: clients/elixir/lib/cliptown_client.ex
defmodule ClipTown.Client do
  @moduledoc "ClipTown HTTP transport."
  defstruct [:base_url, :token, timeout: 30_000]

  def new(base_url, opts \\ []) do
    uri = URI.parse(base_url)
    unless uri.scheme in ["http", "https"] and is_binary(uri.host) and is_nil(uri.userinfo), do: raise(ArgumentError, "base_url must be credential-free absolute HTTP(S)")
    %__MODULE__{base_url: String.trim_trailing(base_url, "/"), token: opts[:token], timeout: opts[:timeout] || 30_000}
  end

  def request(client, method, path, body \\ nil) do
    :inets.start(); :ssl.start()
    url = String.to_charlist(client.base_url <> "/" <> String.trim_leading(path, "/"))
    headers = [{'accept', 'application/json'}] |> authorize(client.token)
    request = if is_nil(body), do: {url, headers}, else: {url, headers, 'application/json', Jason.encode!(body)}
    case :httpc.request(method, request, [timeout: client.timeout, autoredirect: false], body_format: :binary) do
      {:ok, {{_, status, _}, response_headers, response}} when status in 200..299 -> {:ok, status, response_headers, response}
      {:ok, {{_, status, _}, _, response}} -> {:error, status, response}
      {:error, reason} -> {:error, reason}
    end
  end

  defp authorize(headers, nil), do: headers
  defp authorize(headers, ""), do: headers
  defp authorize(headers, token), do: [{'authorization', String.to_charlist("Bearer " <> token)} | headers]
end

File: clients/erlang/src/cliptown_client.erl
-module(cliptown_client).
-export([request/5]).

request(Base0, Token0, Method, Path0, Body) ->
    application:ensure_all_started(inets),
    application:ensure_all_started(ssl),
    Base = to_list(Base0), Path = to_list(Path0), Token = to_list(Token0),
    Url = string:trim(Base, trailing, "/") ++ "/" ++ string:trim(Path, leading, "/"),
    Headers0 = [{"accept", "application/json"}],
    Headers = case Token of [] -> Headers0; _ -> [{"authorization", "Bearer " ++ Token} | Headers0] end,
    Request = case Body of undefined -> {Url, Headers}; _ -> {Url, Headers, "application/json", jsx:encode(Body)} end,
    case httpc:request(Method, Request, [{autoredirect, false}], [{body_format, binary}]) of
        {ok, {{_, Status, _}, ResponseHeaders, Response}} when Status >= 200, Status < 300 -> {ok, Status, ResponseHeaders, Response};
        {ok, {{_, Status, _}, _, Response}} -> {error, Status, Response};
        {error, Reason} -> {error, Reason}
    end.

to_list(Value) when is_binary(Value) -> binary_to_list(Value);
to_list(Value) when is_list(Value) -> Value.
